Fix #41 - Add a game timer

mgoimam opened the pull request:
- Add a 'ScoreValue' object to track the different scored
  properties of a game.
- Add a 'GameMod' interface to represent the different
  add-on behaviors that can be plugged on the base game.
- Add the 'TimedGame' mod that ends the game after 3
  minutes and displays the remaining time.
- Simplified the display logic of the score values.

mgoimam committed:
Fix #41 - Add a game timer

- Add a 'ScoreValue' object to track the different scored
  properties of a game.
- Add a 'GameMod' interface to represent the different
  add-on behaviors that can be plugged on the base game.
- Add the 'TimedGame' mod that ends the game after 3
  minutes and displays the remaining time.
- Simplified the display logic of the score values.

scjody commented:
I like the `GameMod` interface... but looks like you're missing `TimedGame`.  Do you need to `git add` it?

mgoimam committed:
Actually add the GameMod file

scjody commented:
Looks good now - :shipit: 

Note that for issue #41 we also want a way to vary the time before play starts....

scjody commented:
Also, can you please change the time display to MM:SS?

mgoimam committed:
Add time score values

- Move ScoreValue to own file.
- Add a 'toString' method on score values.
- Add 'int' and 'time' (in seconds) score value
  specializations.

mgoimam committed:
Extract menu functions to own class

- Introduce an 'InputHandler' interface to allow different components to
  handle the inputs.
- Add a 'paused' state to the tetris game during which the menu will be
  shown and the game will stop being updated.
- Rename the new game command to a start command since it now handles more
  than starting a new game.

mgoimam committed:
Game duration is now editable

- Display the configured game duration in the menu before game creation
  and allow changing it.
- When game duration is 0, the timed game mode is not instanciated at all
  (infinite mode).
- Add a new 'up' command which is mapped to different keys based on input
  to allow changing the configuration settings.

scjody commented:
Looks good, thanks! :shipit:

Sorry for the delay in reviewing this - I didn't notice that you'd added more commits.  Note that GitHub doesn't notify when you push, only when you comment, so it's a good idea to comment when something is ready to be reviewed again :smile_cat:
